proof/access-control/AARCH64/ArchAccess.thy

```diff
@@ -0,0 +1,363 @@
+   \<union> (if AllowRead \<in> r then {Read} else {})"
+
+definition arch_cap_auth_conferred where
+  "arch_cap_auth_conferred arch_cap \<equiv>
+     (if is_FrameCap arch_cap then vspace_cap_rights_to_auth (acap_rights arch_cap) else {Control})"
+
+subsection \<open>Generating a policy from the current ASID distribution\<close>
+
+definition pte_ref2 where
+  "pte_ref2 level pte \<equiv> case pte of
+     PagePTE paddr _ _ rights
+       \<Rightarrow> Some (ptrFromPAddr paddr,
+                pt_bits_left level,
+                vspace_cap_rights_to_auth rights)
+   | PageTablePTE ppn
+       \<Rightarrow> Some (ptrFromPAddr (paddr_from_ppn ppn), 0, {Control})
+   | _ \<Rightarrow> None"
+
+definition vs_refs_aux :: "vm_level \<Rightarrow> arch_kernel_obj \<Rightarrow> (obj_ref \<times> obj_ref \<times> aa_type \<times> auth) set"
+  where
+  "vs_refs_aux level \<equiv> \<lambda>ko. case ko of
+     ASIDPool pool \<Rightarrow> (\<lambda>(r,p). (p, ucast r, AASIDPool, Control)) ` graph_of (option_map ap_vspace o pool)
+   | PageTable pt \<Rightarrow> (case pt of
+       VSRootPT pt \<Rightarrow> \<Union>(r,(p, sz, auth)) \<in> graph_of (pte_ref2 level o pt).
+                      (\<lambda>(p, a). (p, ucast r, APageTable VSRootPT_T, a)) ` (ptr_range p sz \<times> auth)
+     | NormalPT pt \<Rightarrow> \<Union>(r,(p, sz, auth)) \<in> graph_of (pte_ref2 level o pt).
+                      (\<lambda>(p, a). (p, ucast r, APageTable NormalPT_T, a)) ` (ptr_range p sz \<times> auth))
+   | _ \<Rightarrow> {}"
+
+definition state_vrefs where
+  "state_vrefs s \<equiv> \<lambda>p.
+     \<Union>{vs_refs_aux lvl ao | lvl ao bot asid vref. vs_lookup_table bot asid vref s = Some (lvl, p)
+                                                 \<and> vspace_objs_of s p = Some ao \<and> vref \<in> user_region}"
+
+lemma state_vrefsD:
+  "\<lbrakk> vs_lookup_table level asid vref s = Some (lvl, p);
+     vspace_objs_of s p = Some ao; vref \<in> user_region; x \<in> vs_refs_aux lvl ao \<rbrakk>
+     \<Longrightarrow> x \<in> state_vrefs s p"
+  unfolding state_vrefs_def by fastforce
+
+end
+
+context Arch_p_arch_update_eq begin global_naming AARCH64
+
+interpretation Arch .
+
+lemma state_vrefs[iff]: "state_vrefs (f s) = state_vrefs s"
+  by (simp add: state_vrefs_def pspace)
+
+end
+
+context Arch begin global_naming AARCH64
+
+lemmas state_vrefs_upd =
+  cur_thread_update.state_vrefs
+  cdt_update.state_vrefs
+  irq_node_update_arch.state_vrefs
+  interrupt_update.state_vrefs
+  revokable_update.state_vrefs
+  machine_state_update.state_vrefs
+  more_update.state_vrefs
+
+end
+
+context Arch begin
+
+primrec aobj_ref' where
+  "aobj_ref' (ASIDPoolCap ref _) = {ref}"
+| "aobj_ref' ASIDControlCap = {}"
+| "aobj_ref' (FrameCap ref _ sz _ _) = ptr_range ref (pageBitsForSize sz)"
+| "aobj_ref' (PageTableCap ref _ _) = {ref}"
+| "aobj_ref' (VCPUCap ref) = {ref}"
+
+fun acap_asid' :: "arch_cap \<Rightarrow> asid set" where
```

Comment:
I'm not sure I understand the role of `aobj_ref'` and `acap_asid'`. Are there non-`'` versions? Why are they not what we want in access?

Reply:
The ASpec versions don't return sets, which we do want here.

Reply:
maybe `_set` instead of `'` or a comment then?

Reply:
This is a quirk of access control. I can address this in a later refactor, but it will touch all architectures